be/recovery.c

```diff
@@ -245,16 +245,27 @@ M0_INTERNAL int m0_be_recovery_run(struct m0_be_recovery *rvr)
 		rc = be_recovery_log_record_iter_new(&iter);
 		rc = rc ?: m0_be_log_record_prev(log, prev, iter);
 		if (rc == 0) {
-			log_record_iter_tlink_init_at(iter, &rvr->brec_iters);
-			prev = iter;
-			M0_LOG(M0_DEBUG, "Backward movement : prev header "
-					  BFLRH_F, BFLRH_P(&(prev->lri_header)));
+			if (iter->lri_header.lrh_serial_num == log_hdr.flh_serial_num) {
+				log_record_iter_tlink_init_at(iter, &rvr->brec_iters);
+				prev = iter;
+				M0_LOG(M0_ERROR, "record header=%"PRIu64"log header=%"PRIu64,
+				       prev->lri_header.lrh_serial_num, log_hdr.flh_serial_num);
+				M0_LOG(M0_DEBUG, "Backward movement : prev header "
+						 BFLRH_F, BFLRH_P(&(prev->lri_header)));
+			}
+			else {
+				M0_LOG(M0_ERROR, "log header and record header serial num do not match"
+						 "record header=%"PRIu64"log header=%"PRIu64,
+						 prev->lri_header.lrh_serial_num, log_hdr.flh_serial_num);
+				rc = -EINVAL;
+				break;
+			}
 		} else
 			be_recovery_log_record_iter_destroy(iter);
 	}
 	M0_ASSERT(ergo(rc == 0, prev->lri_header.lrh_pos == last_discarded));
 
-	if (rc != 0){
+	if (rc != 0 && rc != -EINVAL){
```

Comment:
If the serial number mismatch, rc will be -EINVAL. That will not be considered as an error, and we will proceed?
If so, the rc should be reset to 0 ?

Reply:
done.